### PR TITLE
feat(workload): Add prototype support for workloads.

### PR DESCRIPTION
**Pull request by simbuerg**
This converts our implementation of workloads in the form of 'compile' and
'run_tests' into a generic support for workload functions.

The initial implementation aims for feature parity with our current behavior.
Therefore, we only add a 'Phase' enum to determine the execution time of the
workload.

Workloads are executed in insertion order, which is the order of definition in
the project.

A new workload is added by decorating a function with the 'phase' class
function.

This will be a breaking change.

CU-1u9upxa

### CODE REVIEW

**Comment by simbuerg** [comment]
This still misses a sane way to support old ``compile`` and ``run_tests`` definitions properly.

**Comment by simbuerg** [comment]
Still missing:

- Remove old draft implementation of workloads.
- More tests for Command/ProjectCommand, e.g. output_param
- SourceRoot detection without relying on source_of / accept relative dirs as relative to builddir.

File: benchbuild/project.py [hunk 1]
```diff
@@ -147,8 +119,57 @@ def active_variants(self) -> tp.List[VariantContext]:
...
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        if hasattr(cls, 'run_tests'):
+            f_run_tests = run.in_builddir()(run.store_config(cls.run_tests))
+            setattr(cls, 'run_tests', f_run_tests)
+
+        if hasattr(cls, 'compile'):
+            f_compile = run.in_builddir()(run.store_config(cls.run_tests))
```
Comment (vulder) [inline benchbuild/project.py:157]:
```suggestion
            f_compile = run.in_builddir()(run.store_config(cls.compile))
```

**Review by vulder — CHANGES_REQUESTED** [review]
I'm still working on the full review, I just wanted to note a few things.

File: benchbuild/command.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+
+    def __str__(self) -> str:
+        return str(self.command)
+
+
+WorkloadIndex = tp.MutableMapping[WorkloadSet, tp.List[Command]]
+
+
+def unwrap(
+    index: WorkloadIndex, project: 'benchbuild.project.Project'
```
Comment (vulder) [inline benchbuild/command.py:460]: Why do we need to pass a index here? Can unwrap be used on another index?
Normally, `project.workloads` returns the index, which could be called inside unwrap, so we could only pass project and not both.

File: benchbuild/command.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+        return f"WorkloadSet({{{repr_str}}})"
+
+    def unwrap(self, project: "benchbuild.project.Project") -> "WorkloadSet":
+        """Implement the `SupportsUnwrap` protocol."""
+        del project
+        return self
+
+
+class OnlyIn:
```
Comment (vulder) [inline benchbuild/command.py:280]: works like a charm :+1:

**Pull request merged**